external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -335,36 +336,32 @@ private[sql] class AvroDeserializer(
       avroPath: Seq[String],
       catalystPath: Seq[String],
       applyFilters: Int => Boolean): (CatalystDataUpdater, GenericRecord) => Boolean = {
-    val validFieldIndexes = ArrayBuffer.empty[Int]
-    val fieldWriters = ArrayBuffer.empty[(CatalystDataUpdater, Any) => Unit]
-
-    val avroSchemaHelper = new AvroUtils.AvroSchemaHelper(avroType, avroPath)
-    val length = catalystType.length
-    var i = 0
-    while (i < length) {
-      val catalystField = catalystType.fields(i)
-      avroSchemaHelper.getFieldByName(catalystField.name) match {
-        case Some(avroField) =>
-          validFieldIndexes += avroField.pos()
 
+    val avroSchemaHelper =
+      new AvroUtils.AvroSchemaHelper(avroType, catalystType, avroPath, positionalFieldMatch)
+
+    avroSchemaHelper.getCatalystFieldsWithoutMatch.filterNot(_.nullable) match {
```

Comment:
Why do we need this step which goes over the two schemas? We can just check the length of the two schemas is the same and every Avro field has a corresponding catalyst field in the loop below.

Reply:
On the `AvroSerializer`, we could just check the schema lengths, but then when there is a mismatch, you don't get a very helpful error message: "Expected 7 fields but got 6" is not nearly as helpful as "Avro schema is missing field 'foo'". I think it's worth the extra complexity here to have more useful error messages, and I think this argument will be strengthened by the `AvroDeserializer` discussion below.

On the deserializer side, checking the length is not sufficient. This is because we actually allow Avro fields which don't have a matched Catalyst field (note that we don't bother calling `avroSchemaHelper.getAvroFieldsWithoutMatch` in the deserializer), treating this as projection. Similarly, we allow Catalyst fields which don't have a matching Avro field, as long as the Catalyst field is nullable, hence the `filterNot(_.nullable)` call. Neither of these things would be possible if we just checked the schema length. Note that even before this PR, we only perform a check on the schema lengths in the serializer, not the deserializer.

Given that it's actually necessary for us to implement this logic on the deserializer side, I think it makes sense to also leverage it on the serializer side to provide better error messages.

Reply:
Shall we have this improvement in another PR? And we can have test cases of incompatible schemas for both by-name and by-position matching. 
Currently I find this PR a bit complicated and there is not test cases for this.
Supporting the position matching in this one is enough.

Reply:
Sure, reverted, thanks for the input!